src/background/modules/ContextMenu.js

```diff
@@ -0,0 +1,169 @@
+import * as UnicodeTransformationHandler from "/common/modules/UnicodeTransformationHandler.js"
+import * as AddonSettings from "/common/modules/AddonSettings/AddonSettings.js";
+import * as BrowserCommunication from "/common/modules/BrowserCommunication/BrowserCommunication.js";
+import { isMobile } from "/common/modules/MobileHelper.js";
+
+import { COMMUNICATION_MESSAGE_TYPE } from "/common/modules/data/BrowserCommunicationTypes.js";
+import { menuStructure, SEPARATOR_ID, TRANSFORMATION_TYPE } from "/common/modules/data/Fonts.js";
+
+const menus = browser.menus || browser.contextMenus; // fallback for Thunderbird
+const PREVIEW_STRING_CUT_LENGTH = 1000; // a setting that may improve performance by not calculating invisible parts of the context menu
+
+let lastCachedUnicodeFontSettings = null;
+
+/**
+ * Handle context menu click.
+ *
+ * @param {Object} info
+ * @param {Object} tab
+ * @returns {void}
+ * @throws {Error}
+ */
+function handleMenuChoosen(info, tab) {
+    let text = info.selectionText;
+
+    if (!text) {
+        return;
+    }
+
+    text = text.normalize();
+    const menuItem = info.menuItemId;
+    const output = UnicodeTransformationHandler.transformText(text, menuItem);
+
+    browser.tabs.executeScript(tab.id, {
+        code: `insertIntoPage("${output}");`,
+        frameId: info.frameId
+    });
+}
+
+/**
+ * Handle context menu click.
+ *
+ * @param {Object} info
+ * @param {Object} tab
+ * @returns {void}
+ * @throws {Error}
+ */
+async function handleMenuShown(info) {
+    let text = info.selectionText || "";
+
+    // do not show menu entry when no text is selected
+    if (!text) {
+        await menus.removeAll();
+        return menus.refresh();
+    }
+    // shorten preview text as it may not be shown anyway
+    if (text.length > PREVIEW_STRING_CUT_LENGTH) {
+        // to be sure, we append … anyway, in case some strange OS has a tooltip for context menus or so
+        text = `${text.substr(0, PREVIEW_STRING_CUT_LENGTH)}…`;
+    }
+    text = text.normalize();
+
+    const menuIsShown = info.menuIds.length > 0;
+    if (!lastCachedUnicodeFontSettings.livePreview) {
+        if (menuIsShown) {
+            return;
+        }
+
+        // continue re-creating deleted menu, but without any example text
+        text = null;
+    }
+
+    buildMenu(lastCachedUnicodeFontSettings, text, menuIsShown);
+
+    menus.refresh();
+}
+
+/**
+ * Apply (new) menu item settings by (re)creating or updating/refreshing the context menu.
+ *
+ * @param {Object} unicodeFontSettings
+ * @param {string?} [exampleText=null]
+ * @param {bool?} [refreshMenu=false]
+ * @returns {void}
+ */
+function buildMenu(unicodeFontSettings, exampleText = null, refreshMenu = false) {
+    var addedEntries = false;
+    for (const transformationId of menuStructure) {
+        if (transformationId === SEPARATOR_ID) {
+            if (!addedEntries || refreshMenu) {
```

Comment:
Damn, no, thanks. This definitively needs fixing.
I thought this was an easy and elegant solution, but damn… I missed that case.

Note how I did it in the previous PR before, with filtering the array with functional stuff of JS before iterating it. Still thinking on how to best implement it…

Reply:
Do you have ideas for a nice and clean solution?

Reply:
Hmm... The only thing I can think of would be to update the `menuStructure` array to be a nested array, so that each of the two options are separated. Something like this:
```js
export const menuStructure = Object.freeze([
    [
        `${FONT_ID_PREFIX}Superscript`,
        `${FONT_ID_PREFIX}SmallCaps`,
        ...
    ],
    [
        `${CASE_ID_PREFIX}Lowercase`,
        `${CASE_ID_PREFIX}Uppercase`,
        ...
    ]
]);
```
Then, update the for loop to be two nested for loops. Something like this:
```js
let addedEntries = false;
const settings = [unicodeFontSettings.changeFont, unicodeFontSettings.changeCase];
for (let i = 0; i < menuStructure.length; ++i) {
    if (!settings[i]) {
        continue;
    }
    for (let j = 0; j < menuStructure[i].length; ++j) {
        if (menuStructure[i][j] === SEPARATOR_ID || (addedEntries && j === 0)) {
            if (!addedEntries || refreshMenu) {
```
Not sure if this is a "clean solution", but it should work at least.

Reply:
should be fixed now

Reply:
This should be fixed now, though I did use an object of arrays.